scripts/frame/main.js

```diff
@@ -52,115 +79,163 @@ lychee.html = function (literalSections, ...substs) {
 	return result;
 };
 
+/**
+ * @returns {string} - either `"touchend"` or `"click"`
+ */
 lychee.getEventName = function () {
-	let touchendSupport =
+	const touchendSupport =
 		/Android|iPhone|iPad|iPod/i.test(navigator.userAgent || navigator.vendor || window.opera) && "ontouchend" in document.documentElement;
 	return touchendSupport === true ? "touchend" : "click";
 };
 
 // Sub-implementation of lychee -------------------------------------------------------------- //
 
-let frame = {
+const frame = {
+	/** @type {number} */
 	refresh: 30000,
+	/** @type {?Photo} */
+	photo: null,
 };
 
+/**
+ * @returns {void}
+ */
 frame.start_blur = function () {
-	let img = document.getElementById("background");
-	let canvas = document.getElementById("background_canvas");
+	const img = document.getElementById("background");
+	const canvas = document.getElementById("background_canvas");
 	StackBlur.image(img, canvas, 20);
 	canvas.style.width = "100%";
 	canvas.style.height = "100%";
 };
 
+/**
+ * @returns {void}
+ */
 frame.next = function () {
 	$("body").removeClass("loaded");
 	setTimeout(function () {
 		frame.refreshPicture();
 	}, 1000);
 };
 
+/**
+ * @returns {void}
+ */
 frame.refreshPicture = function () {
-	api.post("Photo::getRandom", {}, function (data) {
-		if (!data.url && (data.sizeVariants === null || data.sizeVariants.medium === null)) {
-			console.log("URL not found");
-		}
-		if (data.sizeVariants.thumb === null) console.log("Thumb not found");
-
-		$("#background").attr("src", data.sizeVariants.thumb.url);
-
-		let srcset = "";
-		let src = "";
-		this.frame.photo = null;
-		if (data.sizeVariants.medium !== null) {
-			src = data.sizeVariants.medium.url;
-
-			if (data.sizeVariants.medium2x !== null) {
-				srcset = `${data.sizeVariants.medium.url} ${data.sizeVariants.medium.width}w, ${data.sizeVariants.medium2x.url} ${data.sizeVariants.medium2x.width}w`;
-				// We use it in the resize callback.
-				this.frame.photo = data;
+	api.post(
+		"Photo::getRandom",
+		{},
+		/** @param {Photo} data */
+		function (data) {
+			// TODO: My IDE complains that this condition is always false, because each Photo has at least a thumbnail
+			if (data.size_variants === null || (data.size_variants.original === null && data.size_variants.medium === null)) {
+				console.log("URL not found");
+			}
+			// TODO My IDE complains that this condition is always false. Is this legacy?
```

Comment:
Well before your refactoring some pictures did not have thumbs.

Reply:
Maybe there was I time when not all pictures had thumbnails but that must have been changed long before I refactored it. When I refactored the code, thumbs were always created even if the original was smaller than 200px (400px), resp. So I assume this condition is legacy and can be re-factored.

Reply:
See my comments from yesterday: videos may be missing thumbnails.